internal/service/service.go

```diff
@@ -0,0 +1,98 @@
+package service
+
+import (
+	"context"
+	"fmt"
+
+	m "github.com/PDOK/uptime-operator/internal/model"
+	"github.com/PDOK/uptime-operator/internal/service/providers"
+	"sigs.k8s.io/controller-runtime/pkg/log"
+)
+
+type UptimeCheckOption func(*UptimeCheckService) *UptimeCheckService
+
+type UptimeCheckService struct {
+	provider UptimeProvider
+	slack    *Slack
+}
+
+func New(options ...UptimeCheckOption) *UptimeCheckService {
+	service := &UptimeCheckService{}
+	for _, option := range options {
+		service = option(service)
+	}
+	return service
+}
+
+func WithProvider(provider UptimeProvider) UptimeCheckOption {
+	return func(service *UptimeCheckService) *UptimeCheckService {
+		service.provider = provider
+		return service
+	}
+}
+
+func WithProviderName(provider string) UptimeCheckOption {
+	return func(service *UptimeCheckService) *UptimeCheckService {
+		switch provider { //nolint:gocritic
+		case "mock":
+			service.provider = providers.NewMockUptimeProvider()
+			// TODO add new case(s) for actual uptime monitoring SaaS providers
+		}
+		return service
+	}
+}
+
+func WithSlack(slackToken string, slackChannel string) UptimeCheckOption {
+	return func(service *UptimeCheckService) *UptimeCheckService {
+		if slackToken != "" && slackChannel != "" {
+			service.slack = NewSlack(slackToken, slackChannel)
+		}
+		return service
+	}
+}
+
+func (r *UptimeCheckService) Mutate(ctx context.Context, mutation m.Mutation, ingressName string, annotations map[string]string) {
```

Comment:
als het mutaten niet lukt, gaat er geen error terug. dan lukt de reconciliation. is dat gewenst? als je wel een error terug stuurt, komt de reconciliation weer in de queue, al dan niet met vertraging. nu komt er "niet eens" een status (is ook niet onze resource). voor deleten / finalizer is dan misschien wel weer fijn dat het door kan. anders staat iemand te wachten op de finalizer tenzij force. misschien is er een mechanisme te bedenken om uiteindelijk toch de boel opgeruimd te krijgen. een tussen-cr misschien. gaat wat ver wellicht. of afhankelijk van de fout wel of niet terug te kaatsen (transient error of niet. network error retryen, maar resource bestaat niet, niet)

Reply:
(voor nu sowieso prima zo)

Reply:
als mutaten mislukt gaat er (tenzij harde crash maar dan zie je wel dat de operator in crash-loop komt) een slack message eruit. Zou het (nu) niet complexer maken. Extra CRs etc klinkt wel als overkill. En daarnaast de resource is idd niet van deze operator.

Reply:
o ja, slack message, niet aan gedacht, dat is top

Reply:
Was jouw idee 👍